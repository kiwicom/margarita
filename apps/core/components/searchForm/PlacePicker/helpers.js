// @flow

import { type Location } from '../../../scenes/search/SearchContext';

export type OptionType = {
  +id: string | number,
  +locationId?: string,
  +label: string,
  +type: 'localization' | 'destination' | 'airplane' | 'bus' | 'train',
};

export function mapOptionToLocation(option: OptionType): Location {
  return {
    id: option.id,
    name: option.label,
    locationId: option.locationId,
    type: option.type,
  };
}

export function convertLocationTypeToOptionType(type: ?string) {
  switch (type) {
    case 'AIRPORT':
      return 'airplane';
    case 'BUS_STATION':
      return 'bus';
    case 'STATION':
      return 'train';
    default:
      return 'destination';
  }
}

export function filterOptions(options: OptionType[], selected: OptionType[]) {
  const selectedIDs = selected.map(({ id }) => id);

  return options.reduce((acc = [], option) => {
    // check if the option is selected, if yes filter it out
    if (selectedIDs.includes(option.id)) {
      return acc;
    }
    return [...acc, option];
  }, []);
}
