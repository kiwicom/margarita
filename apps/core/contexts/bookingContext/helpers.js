// @flow

import type { PassengerType } from './BookingContext';

const PASSENGER_FIELDS = [
  'name',
  'lastName',
  'gender',
  'nationality',
  'dateOfBirth',
  'id',
  'insurance',
  'bags',
  'visaRequired',
];

export function createPassengers(count: number): Array<PassengerType> {
  const passengers = Array.from({ length: count });

  return passengers.map<Object>(() =>
    PASSENGER_FIELDS.reduce((acc, field) => ({ ...acc, [field]: null }), {}),
  );
}
