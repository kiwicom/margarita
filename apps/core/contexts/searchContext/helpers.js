// @flow

import qs from 'qs';

import type { Location } from './SearchContextTypes';

type ParserType = 'Location' | 'Date' | 'number' | 'string' | 'boolean';

const PARSER_CONFIG = {
  travelFrom: 'Location',
  travelTo: 'Location',
  travelFromName: 'string',
  travelToName: 'string',
  sortBy: 'string', // @TODO create sortParser
  limit: 'number',
  adults: 'number',
  infants: 'number',
  dateFrom: 'Date',
  dateTo: 'Date',
  returnDateFrom: 'Date',
  returnDateTo: 'Date',
  bookingToken: 'string',
  tripType: 'string',
  nightsInDestinationFrom: 'number',
  nightsInDestinationTo: 'number',
  isNightsInDestinationSelected: 'boolean',
};

export function parseURLqueryToState(query: Object) {
  const queryKeys = Object.keys(query);

  return queryKeys.reduce((acc, key) => {
    const parserType = PARSER_CONFIG[key];
    if (parserType) {
      const parser = getParser(parserType);

      return { [key]: parser(query[key]), ...acc };
    }
    /* eslint-disable-next-line no-console */
    console.warn(`Unexpected URL parameter "${key}" have been detected`);
    return acc;
  }, {});
}

function getParser(parserType: ParserType) {
  switch (parserType) {
    case 'Location': {
      return locationParser;
    }
    case 'Date': {
      return dateParser;
    }
    case 'string': {
      return stringParser;
    }
    case 'number': {
      return numberParser;
    }
    case 'boolean': {
      return booleanParser;
    }
    default: {
      return (value: any) => value;
    }
  }
}

export function locationParser(query: string) {
  // @TODO Location validation
  const locationsObject = qs.parse(query);
  const locationsArray = Object.keys(locationsObject).map<Location>(
    key => locationsObject[key],
  );
  return locationsArray;
}

function dateParser(date: string) {
  // @TODO input validation
  return new Date(date);
}

function numberParser(number: number | string) {
  // @TODO input validation
  return parseInt(number, 10);
}

function stringParser(string: string) {
  // @TODO input validation
  return string;
}

function booleanParser(boolean: boolean) {
  // @TODO input validation
  return boolean;
}
