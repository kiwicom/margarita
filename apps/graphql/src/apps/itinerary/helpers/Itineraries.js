// @flow

import * as DateFNS from 'date-fns';
import { head, last } from 'ramda';
import { fromGlobalId } from '@kiwicom/graphql-global-id';

import type { ApiRouteItem, Sector, Segment } from '../Itinerary';
import type { RouteStop } from '../../booking/Booking';

export const differenceInMinutes = (
  from: ?(string | number),
  to: ?(string | number),
) => {
  if (from == null || to == null) {
    return null;
  }
  const parseNonSpecificDate = date =>
    typeof date === 'string'
      ? DateFNS.parseISO(date)
      : DateFNS.fromUnixTime(date);

  return parseInt(
    DateFNS.differenceInMinutes(
      parseNonSpecificDate(to),
      parseNonSpecificDate(from),
    ),
    10,
  );
};

export const mapToOpaqueID = (input: any) => {
  const parsedOriginIds: string[] = input.itinerary.origin.ids.map(id =>
    fromGlobalId(id),
  );
  const parsedDestinationIds: string[] = input.itinerary.destination.ids.map(
    id => fromGlobalId(id),
  );

  return {
    ...input,
    itinerary: {
      ...input.itinerary,
      origin: {
        ids: parsedOriginIds,
      },
      destination: {
        ids: parsedDestinationIds,
      },
    },
  };
};

export const mapVehicle = (type: ?string, uniqueNo: ?string) => ({
  type: type ?? null,
  uniqueNo: uniqueNo ?? null,
});

export const mapTransporter = (transporter: ?string) => ({
  name: transporter ?? null,
});

export const getItineraryType = (routes: ?Array<Array<string>>) => {
  if (routes == null) {
    return null;
  }
  if (routes.length === 1) {
    return 'oneway';
  }
  if (routes.length === 2) {
    return 'return';
  }
  return null;
};

/**
 * This function converts route and routes data from REST API to well-structured Sectors with Segments.
 * Sector = The whole part from source to destination, e.g Oslo -> Prague. A return trip will have 2 sectors
 * and 1 way trip will have 1 sector. A multicity trip could have 2 or more sectors.
 * Segment = A part of the sector. Segments could be Oslo -> Warzaw, Warzaw -> Prague
 */
export const mapSectors = (
  routesList: ?(ApiRouteItem[]),
  routeCodes: ?Array<string[]>,
): ?Array<Sector> => {
  if (routesList == null || routeCodes == null) {
    return null;
  }
  const sectors = [];
  let currentSectorIndex = 0;
  let currentRouteCode = head(routeCodes) ?? [];

  let { currentSector, currentArrival } = sanitizeSector(
    routesList,
    currentRouteCode,
  );

  sectors[currentSectorIndex] = currentSector;

  routesList.forEach((route: ApiRouteItem) => {
    const segment = sanitizeSegment(route);

    if (sectors[currentSectorIndex].segments) {
      sectors[currentSectorIndex].segments.push(segment);
    }

    if (route.flyTo === currentArrival?.flyTo) {
      currentSectorIndex++;
      // $FlowExpectedError: We already tested that routeCodes != null
      currentRouteCode = routeCodes[currentSectorIndex];

      if (currentRouteCode == null) {
        return;
      }
      // $FlowExpectedError: We already tested that routesList != null
      const nextSector = sanitizeSector(routesList, currentRouteCode);
      currentArrival = nextSector.currentArrival;
      sectors[currentSectorIndex] = nextSector.currentSector;
    }
  });

  return sectors;
};

const sanitizeSector = (
  routesList: ApiRouteItem[],
  currentRouteCode: string[],
) => {
  const currentRouteArrivalCode = last(currentRouteCode) ?? '';
  const currentRouteDepartureCode = head(currentRouteCode) ?? '';

  const currentDeparture: ?ApiRouteItem = routesList.find(
    route => route.flyFrom === currentRouteDepartureCode,
  );

  const currentArrival: ?ApiRouteItem = routesList.find(
    route => route.flyTo === currentRouteArrivalCode,
  );

  const departure = sanitizeRouteStop(
    apiRouteItemToDeparture(currentDeparture),
  );
  const arrival = sanitizeRouteStop(apiRouteItemToArrival(currentArrival));

  const currentSector = {
    departure,
    arrival,
    arrivalTime: {
      local: arrival.time?.local,
      utc: arrival.time?.utc,
    },
    departureTime: {
      local: departure.time?.local,
      utc: departure.time?.utc,
    },
    duration: differenceInMinutes(departure.time?.utc, arrival.time?.utc),
    destination: mapLocation(currentArrival?.flyTo, currentArrival?.cityTo),
    origin: mapLocation(currentDeparture?.flyFrom, currentDeparture?.cityFrom),
    segments: [],
    stopoverDuration: 0,
  };

  return {
    currentArrival,
    currentSector,
  };
};

const apiRouteItemToArrival = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityTo,
  code: routeItem?.flyTo,
  time: {
    utc: routeItem?.utc_arrival,
    local: routeItem?.local_arrival,
  },
});

const apiRouteItemToDeparture = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityFrom,
  code: routeItem?.flyFrom,
  time: {
    utc: routeItem?.utc_departure,
    local: routeItem?.local_departure,
  },
});

const sanitizeRouteStop = ({
  cityName,
  code,
  time,
}: {|
  +cityName: ?string,
  +code: ?string,
  +time: {| +utc: ?string, +local: ?string |},
|}): RouteStop => ({
  cityName,
  cityId: null,
  time,
  code,
});

const sanitizeSegment = (segment: ?ApiRouteItem): Segment => {
  return {
    duration: differenceInMinutes(segment?.utc_departure, segment?.utc_arrival),
    id: segment?.id,
    transporter: mapTransporter(segment?.airline),
    vehicle: mapVehicle(segment?.vehicle_type, String(segment?.flight_no)),
    departure: sanitizeRouteStop(apiRouteItemToDeparture(segment)),
    arrival: sanitizeRouteStop(apiRouteItemToArrival(segment)),
  };
};

export const mapDate = (local: ?string, utc: ?string) => {
  const localDate = local && DateFNS.parseISO(local).toISOString();
  const utcDate = utc && DateFNS.parseISO(utc).toISOString();
  return {
    local: localDate ?? null,
    utc: utcDate ?? null,
  };
};

export const mapLocationArea = (
  code: ?string,
  locationId: ?string,
  name: ?string,
  slug: ?string,
  flagURL: ?string,
) => ({
  id: locationId ?? null,
  code: code ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  slug: slug ?? null,
  flagURL: flagURL ?? null,
});

export const mapLocation = (
  locationId: ?string,
  name: ?string,
  countryName: ?string,
  countryCode: ?string,
  slug: ?string,
) => ({
  id: locationId ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  timezone: 'UTC+1',
  country: mapLocationArea(countryCode, countryCode, countryName),
  slug: slug ?? null,
  city: null,
  type: null,
});
