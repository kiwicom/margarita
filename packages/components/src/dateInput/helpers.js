// @flow

export const composeDateFromStrings = (
  day: string,
  month: string,
  year: string,
): Date => {
  return new Date(parseFloat(year), parseFloat(month), parseFloat(day));
};
