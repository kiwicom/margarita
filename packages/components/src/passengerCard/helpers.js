// @flow

type Gender = 'female' | 'male' | 'other';

export function getPassengerTitle(gender?: Gender) {
  switch (gender) {
    case 'male':
      return 'Mr.';
    case 'female':
      return 'Ms.';
    default:
      return '';
  }
}
