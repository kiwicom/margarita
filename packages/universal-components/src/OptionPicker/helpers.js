// @flow

import type { OptionTypeInterface } from './OptionPickerTypes';

export function getOptionsIds(options: OptionTypeInterface[]) {
  return options.reduce((result, option) => {
    let ids = [option.id];
    if (option.subOptions) {
      const subOptions = option.subOptions.map(({ id }) => id);
      ids = [...ids, ...subOptions];
    }

    return [...result, ...ids];
  }, []);
}

export function areSelectedOptionsChanged(
  optionsA: OptionTypeInterface[],
  optionsB: OptionTypeInterface[],
) {
  const optionsIdsB = getOptionsIds(optionsB);
  const optionsIdsA = getOptionsIds(optionsA);

  return (
    optionsIdsA.length !== optionsIdsB.length ||
    !optionsIdsA.every(option => optionsIdsB.includes(option))
  );
}
