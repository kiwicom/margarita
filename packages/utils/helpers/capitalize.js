// @flow

export const capitalize = (text: string) => {
  if (text.length > 0) {
    return text[0].toUpperCase() + text.slice(1);
  }
  return text;
};
