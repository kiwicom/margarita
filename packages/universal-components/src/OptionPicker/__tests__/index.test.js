// @flow

import { getOptionsIds, areSelectedOptionsChanged } from '../helpers';
import mockedPlaces from '../__mocks__/places';

describe('OptionPicker: getOptionsIds helper function', () => {
  test('extract the places Ids to the array', () => {
    expect(getOptionsIds(mockedPlaces)).toMatchSnapshot();
  });
  test('empty array as parameter', () => {
    expect(getOptionsIds([])).toEqual([]);
  });
});

describe('OptionPicker: areSelectedOptionsChanged helper function', () => {
  test('should be eual', () => {
    expect(areSelectedOptionsChanged(mockedPlaces, mockedPlaces)).toBeFalsy();
  });
  test('empty array as parameter', () => {
    expect(
      areSelectedOptionsChanged(mockedPlaces, [mockedPlaces[1]]),
    ).toBeTruthy();
  });
});
