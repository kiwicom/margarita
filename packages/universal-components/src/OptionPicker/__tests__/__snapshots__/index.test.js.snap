// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionPicker: getOptionsIds helper function extract the places Ids to the array 1`] = `
Array [
  "prague",
  "train-station",
  "airport",
  "berlin",
  "QPP",
  "QWC",
  "presov",
]
`;
